### DISCUSSION TITLE
How to build Magick.Native-Q16-HDRI-x64.dll from ImageMagick-Windows\VisualMagick?

### DISCUSSION THREAD

**Discussion by swuqi** [ImageMagick/ImageMagick#3098] [Development] [answered]
For security concern, we cannot use some NuGet packages. After I downloaded ImageMagick-Windows package and successfully built it, there is a COM module ImageMagickObject.dll and convert.exe generated in the bin folder. But Magick.NET project only refers to Magick.Native-Q16-HDRI-x64.dll. 
Where can I find the project or instructions to generate Magick.Native-Q16-HDRI-x64.dll from ImageMagickObject.dll or something else?

Thank you for any help.

**Comment by dlemstra** [comment]
You will need the code in the Magick.Native repository of my account. But what is the security issue? The code is build with Github actions.

**Comment by swuqi** [comment]
It’s not about this project’s security issue. The policy about adopting open source code requires to build modules from verifiable source code files instead of downloading DLL or EXE directly.

**Comment by dlemstra** [comment]
You can only use an open source project when you build the DLL inside the project? Or do you have your own NuGet packages?

**Comment by swuqi** [comment]
We don't have NuGet package. We build the DLL used in the project.



-------- Original message --------

**Comment by dlemstra** [comment]
Does this also mean you need to build the Native library yourself? That is not an easy process.

**Comment by swuqi** [comment]
The build of ImageMagick-Windows is successful. What do you mean native library?


-------- Original message --------

**Comment by swuqi** [comment, marked as answer]
Here is my build instructions.

1. Download Magick.Native respository from https://github.com/dlemstra/Magick.Native
2. Extract Magick.Native folder to c:\Projects

3. Copy nuget.exe from "C:\Projects\Magick.NET\tools\windows\nuget.exe" to C:\Projects\Magick.Native\tools\windows

4. Run "C:\Projects\Magick.Native\src\ImageMagick\checkout.cmd" to download ImageMagick package to C:\Projects\Magick.Native\src\ImageMagick as libraries directory.  The libraries directory includes a subset of ImageMagick-Windows (another github package) directories. 

5. run "C:\Projects\Magick.NET\src\Magick.Native\build\Release-Q16-HDRI-x64.cmd"
